refactor/eslint flow support

dlouzan opened the pull request:
- add to repository flow-typed support files (same as verdaccio parent project)
- integrate flow cli into lint script
- modify existing source files to make them pass linting

dlouzan commented:
Builds on top of the auth-cache PR. Mimics flow support in verdaccio

dlouzan commented on .eslintrc.yml:
```diff
@@ -1,268 +1,33 @@
+  ecmaFeatures:
+    impliedStrict: true
+    jsx: true
 rules:
```
The rules are removed intentionally; lots of them conflict with flow support, so for now just enabled basic style checks, more detailed ones can be added later.

juanpicado commented on flow-typed/npm/verdaccio_vx.x.x.js:
```diff
@@ -0,0 +1,367 @@
+// flow-typed signature: 70d4cf81a06984155bd79f63db1449a3
+// flow-typed version: <<STUB>>/verdaccio_v^3.1.1/flow_v0.75.0
```
You should be able to use  https://github.com/verdaccio/flow-types instead.
```
@verdaccio/types
```

juanpicado replied:
Ah, nevermind I just saw it's being used, but ... then I don't get this.

dlouzan replied:
You are right, the `@verdaccio/types` are not needed in the flow-typed directory, the import is enough. I'm pushing a correction in a minute.

juanpicado commented on src/gitlab.js:
```diff
@@ -3,82 +3,127 @@
+  gitlab?: boolean
+}
+
 export default class VerdaccioGitLab {
```
I just shipped `IBasicAuth` interface to help with class structure.
https://github.com/verdaccio/flow-types/blob/master/src/types.js#L345

dlouzan replied:
Ok, but as I understand that commit, that is a non-backwards compatible change, isn't it? It means we have to build the plugin on top of the latest verdaccio version, ?

juanpicado replied:
this project already uses `"@verdaccio/types": "^3.0.0"` so it should works. I hope that's what you are referring for.

dlouzan replied:
Mmm ok maybe I'm not understanding this. In the current code I'm using, `allow_access` and `allow_publish` have the following signatures:

`allow_access(user: { name: string }, _package: VerdaccioGitlabPackageAccess, cb: Callback)`  
`allow_publish(user: { name: string, real_groups: string[] }, _package: VerdaccioGitlabPackageAccess, cb: Callback)`

But in the `@verdaccio/types` you linked, I see this in `verdaccio$IPluginAuth`:

`allow_access(packageName: string, user: string, cb: verdaccio$Callback): void;`  
`allow_publish(packageName: string, user: string, cb: verdaccio$Callback): void;`

Particularly, in the current code I receive an object structure for both the user and package data, whereas in the new interface i just get a string.

Am I missing anything?

juanpicado replied:
`@verdaccio/types@3.3.3` should be right alredy

dlouzan replied:
Just checked, in my code the user is still an object with properties, but the signature in the types just gives me a string, :-?

dlouzan replied:
When I try to implement the interface such as:

```
export default class VerdaccioGitLab implements IPluginAuth {
...
allow_access(user: { name: string }, _package: PackageAccess, cb: Callback) {
...
allow_publish(user: { name: string, real_groups: string[] }, _package: PackageAccess, cb: Callback) {
```

I have the following issues (apart from the mentioned `user: string` that is actually an object):

```
- Property gitlab is missing in verdaccio$PackageAccess [1].
- Cannot get _package.name because property name is missing in verdaccio$PackageAccess [1].
```

Any ideas? I'm actually using those fields in the provided objects, and they do exist. How to reconcile the type definition with what is actually needed?

juanpicado replied:
> Just checked, in my code the user is still an object with properties, but the signature in the types just gives me a string, :-?

omg ... sorry ... I must be asleep. I'll fix it again.

juanpicado replied:
` @verdaccio/types@3.4.1` add a `RemoteUser` and allow SubTypes for `PackageAccess` on  type that looks like

```
declare type verdaccio$RemoteUser = {
  real_groups: Array<string>;
  groups: Array<string>;
  name: string | void;
  error?: string;
}

declare interface verdaccio$IPluginAuth extends verdaccio$IPlugin {
  authenticate(user: string, password: string, cb: verdaccio$Callback): void;
  adduser(user: string, password: string, cb: verdaccio$Callback): void;
  allow_access(user: verdaccio$RemoteUser, pkg: $Subtype<verdaccio$PackageAccess>, cb: verdaccio$Callback): void;
  allow_publish(user: verdaccio$RemoteUser, pkg: $Subtype<verdaccio$PackageAccess>, cb: verdaccio$Callback): void;
}
```

juanpicado replied:
eg: https://github.com/verdaccio/verdaccio/pull/815/commits/cde695a18f444199e8d8e0d6174c3529d366aa39#diff-ac171c117d06952be3c28d8f4fbb871aR64

juanpicado commented:
`"@verdaccio/types": "^3.0.0",` should be a Dev Dependencie ⚠️

juanpicado commented on src/gitlab.js:
```diff
@@ -23,25 +29,47 @@ export default class VerdaccioGitLab {
     this.stuff = stuff;
 
-    this.logger = stuff.logger;
+    if ((this.config.authCache || {}).enabled === false) {
```
IMHO this should be disabled by default. It might confuse users @bufferoverflow whether do have old configuration.

juanpicado replied:
In many cases the will expect that the cache works in multiple processes and will drive to ticket asking why it does not work.

dlouzan replied:
Well, @bufferoverflow can take the final decision, but I think this plugin is not really usable out of testing without a cache, the amount of login calls performed on the upstream gitlab would be huge for any reasonable sized load.

This is untested, but I don't think there will be issues with multiple processes, just that more logins will happen (basically n times more, where n is the number of processes), but each process will keep their own copy of the cache.

I'd be more worried at the moment about the memory usage of each process with a bigger load :-)

bufferoverflow replied:
major version bump can be done after this is in, so using cache by default is fine

bufferoverflow commented:
Awesome collaboration @dlouzan and @juanpicado !

Do we really need `flow-typed` folder within the repo?

juanpicado commented:
@bufferoverflow yes, but not all of them, @dlouzan has included too many, only libraries need it for production should be included. 
![welcome_ _verdaccio-gitlab](https://user-images.githubusercontent.com/558752/42797808-44cd7d52-8991-11e8-8652-07e40c4badf1.png)

juanpicado commented:
`.dockerignore` should be updated excluding `flow-typed/` folder for instance.

dlouzan commented:
@juanpicado @bufferoverflow Latest batch of changes pushed:
- @verdaccio/types is now a dev dep
- Refactored the plugin main class, constructor and methods signatures to comply with the latest @verdaccio/types
- Removed the extra flow-typed entries you marked as dev
- Removed the `files` entry from `package.json` and setup `.npmignore` to ignore everything except `bin/` and `lib/*.js`

As usual, after this is deemed as ok I'll squash all commits together.

bufferoverflow commented:
Just tested locally, it works!

Sehr gut gemacht @dlouzan ! Thanks @juanpicado for your strong collaboration and support on this!

bufferoverflow commented:
very cool!
